From 55a2bfe5be442a67cd834070b942cd0eabd3a82b Mon Sep 17 00:00:00 2001
From: Michael Mann <mmann78@netscape.net>
Date: Tue, 24 Jan 2017 23:06:04 -0500
Subject: [PATCH 140/144] rtmpt: Ensure sequence count is incremented for
 stored fragments

Bug: 13347
Change-Id: I351c80dea8ac7a9f2540b40782b1cc5c0b8fdaed
Reviewed-on: https://code.wireshark.org/review/19777
Reviewed-by: Anders Broman <a.broman58@gmail.com>
(cherry picked from commit ee185445f410b6bc54831ea0923af08cbcd75d00)
Reviewed-on: https://code.wireshark.org/review/19837
Petri-Dish: Michael Mann <mmann78@netscape.net>
Tested-by: Petri Dish Buildbot <buildbot-no-reply@wireshark.org>
(cherry picked from commit 2b3a0909beff8963b390034c594e0b6be6a4e531)
Reviewed-on: https://code.wireshark.org/review/20491
Reviewed-by: Balint Reczey <balint@balintreczey.hu>
---
 epan/dissectors/packet-rtmpt.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/epan/dissectors/packet-rtmpt.c b/epan/dissectors/packet-rtmpt.c
index 24fd20b..9849dd1 100644
--- a/epan/dissectors/packet-rtmpt.c
+++ b/epan/dissectors/packet-rtmpt.c
@@ -1887,6 +1887,7 @@ dissect_rtmpt_common(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, rtmpt_
         guint8  cmd;
         guint32 src;
         int     chunk_size;
+        guint32 save_seq = 0;
 
         rtmpt_frag_t   *tf;
         rtmpt_id_t     *ti;
@@ -1909,8 +1910,9 @@ dissect_rtmpt_common(tvbuff_t *tvb, packet_info *pinfo, proto_tree *tree, rtmpt_
                 wmem_stack_push(packets, 0);
 
                 tp = (rtmpt_packet_t *)wmem_tree_lookup32_le(rconv->packets[cdir], seq+remain-1);
-                while (tp && tp->lastseq >= seq) {
+                while (tp && tp->lastseq >= seq && tp->lastseq >= save_seq) {
                         wmem_stack_push(packets, tp);
+                        save_seq = tp->lastseq+1; /* Ensure sequence is increasing */
                         tp = (rtmpt_packet_t *)wmem_tree_lookup32_le(rconv->packets[cdir], tp->lastseq-1);
                 }
 
-- 
2.1.4

